client/src/main/scala/drt/client/services/handlers/EditUserHandler.scala

```diff
@@ -0,0 +1,47 @@
+package drt.client.services.handlers
+
+import autowire._
+import boopickle.Default._
+import diode.Implicits.runAfterImpl
+import diode._
+import diode.data.{Pending, Pot, Ready}
+import drt.client.actions.Actions._
+import drt.client.logger.log
+import drt.client.services.{AjaxClient, PollDelay}
+import drt.shared.Api
+import drt.shared.KeyCloakApi.KeyCloakGroup
+
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.scalajs.concurrent.JSExecutionContext.Implicits.queue
+
+class EditUserHandler[M](modelRW: ModelRW[M, Pot[Set[KeyCloakGroup]]]) extends LoggingActionHandler(modelRW) {
+  protected def handle: PartialFunction[Any, ActionResult[M]] = {
+    case GetUserGroups(userId) =>
+      log.info(s"Getting groups for $userId")
+      effectOnly(Effect(AjaxClient[Api].getKeyCloakUserGroups(userId).call().map(SetSelectedUserGroups).recoverWith {
+        case _ =>
+          log.error(s"GetUserGroups request failed. Re-requesting after ${PollDelay.recoveryDelay}")
+          Future(RetryActionAfter(GetUserGroups(userId), PollDelay.recoveryDelay))
+      }))
+
+    case SetSelectedUserGroups(groups) =>
+      updated(Ready(groups), Effect(Future(GetKeyCloakUsers)))
+
+    case SaveUserGroups(userId, groupsToAdd, groupsToRemove) =>
+      val futureAddResult: Future[Any] = AjaxClient[Api]
+        .addUserToGroups(userId, groupsToAdd)
```

Comment:
Just a thought - would it be a simpler interface if we had a setGroupsForUser(userID, groups). Might eliminate the need to figure out which groups need adding and removing specifically and reduce the calls from 2 to 1

Reply:
Yeah, you have to do separate requests to add and remove. I could make it remove *all* groups and then add in the new ones, but then we'd lose some meta data around when a user joined a group possibly. Not sure if that matters.